### PR TITLE
Api to trigger a force merge operation on an index

### PR DESCRIPTION
**Pull request by sreekanth-cb**
Introducing a new synchronous ForceMerge api to enable
users conveniently invoke a manual merge operation
on an online scorch index. This api would let advanced
bleve users to compact their index all the way down to a
single file segment based index.

The api takes an optional mergePlanOptions argument which
would be applied for the forceful merge operations, while
the original mergePlanOptions property of the index remains
intact after the forceful merge cycle exits.
In the absense of a mergePlanOptions argument, default policy
assumed would be an aggressive one to target a single file
segment index with a maximum segment size of 1B approx.

The api would trigger the merger's work loop only once
during it's invocation and hence it is the caller's
responsibility to actively monitor the expected
file segments count at the root and trigger any
subsequent forceMerge invocations if needed to
meet the expected segment's count at root.

The caller could also pass a ctx for cancelling
an ongoing forceMerge operation in case of any
runtime anomalies during the merge operation.

The caller should also ensure that there is enough disk
space before invoking this compaction/merge operation
as it needs at least double the disk space
whenever there is a full compaction.

### CODE REVIEW

File: index/scorch/merge.go [hunk 1]
```diff
@@ -158,6 +256,28 @@ func (s *Scorch) planMergeAtSnapshot(ourSnapshot *IndexSnapshot,
...
+		}
+	}
+	defer cleanup()
+
+	// cancel the merge operation on events like the index closure
+	// or upon a user cancel.
+	go func() {
+		select {
+		case <-s.closeCh:
+			cleanup()
```
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index/scorch/merge.go:274]: closeCh is the wrapper signal channel for the index closure s.closeCh and the merge operation's cancelCh from user. 
If either of them happens - then closeCh is used to/called explicitly to cancel the ongoing merge operations.
cleanup() closes closeCh only if it's not already closed, hence there shouldn't be a double close problem.

File: index/scorch/merge.go [hunk 2]
```diff
@@ -106,6 +133,76 @@ OUTER:
...
+	// disk space requirements during a forced merge operation.
+	OverrideNumSnapshotsToKeep bool
+
+	// CancelCh helps in cancelling an ongoing merge operation.
+	CancelCh chan struct{}
+}
+
+// ForceMerge helps users trigger a merge operation on
+// an online scorch index.
+func (s *Scorch) ForceMerge(mr *MergeRequest) error {
```
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index/scorch/merge.go:163]: my bad..will switch to ctx.

[the remark this answers is not in the document]
Comment (mschoch) [reply index/scorch/merge.go:163]: One thing when you do, note there are some best practices around the usage (see context pkg docs for full details)

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx

Further, I think this rule applies to the user facing interface (ie, it should be the first argument to ForceMerge), but that it still could be ok to store that ctx in a struct used internally during the processing of that request...

[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index/scorch/merge.go:163]: thanks for this tip!

File: index_alias_impl.go [hunk 3]
```diff
@@ -44,6 +44,18 @@ func NewIndexAlias(indexes ...Index) *indexAliasImpl {
...
+// Caller must be aware that the results will be
+// inconsistent if there are concurrent Add/Remove
+// operations on the alias.
+func (i *indexAliasImpl) Indexes() []Index {
+	i.mutex.RLock()
+	rv := i.indexes
+	i.mutex.RUnlock()
+	return rv
+}
+
```
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index_alias_impl.go:58]: At the cbft level,I had to invoke the forceMerge on an index alias so that it get triggers only every underlying partitions. And for doing that - i need to perform something similar to the MultiSearch(https://github.com/blevesearch/bleve/blob/master/index_alias_impl.go#L162) in alias.

So, needed an api which exposes the underlying indexes in an alias so that I can scatter-gather any non Search/docCount functionality's/tasks on the indexes at the application level is the motive there.

[the remark this answers is not in the document]
Comment (mschoch) [reply index_alias_impl.go:58]: In that case, I'm not sure why the limitations of this function are not a problem.  Can you actually ensure at the cbft level that no other changes to the alias happen?

Another idea (just thinking out loud), what if instead the caller passed in a function, and we guaranteed to invoke that function on every index in the alias at that time.  I'm not sure that's a whole lot better, but just wondering would that work for you?

**Comment by sreekanth-cb** [comment]
> forceMergeRequestCh 
I like this too.. ^^

Honestly naming is very hard for me. 
To answer - why it was named as "mergerKickCh"  earlier could be due to 2 things.
1. Influenced from all the janitorKick/Planner Kick channels in cbgt..  ;)
2. Felt like a bit safe name as the "final decision as well as the real merging is eventually at the whole discretion of the merger and as we are just kicking the merger for another work loop thought the a name indicating that manoeuver would be nice..
But now I see the flaw in that thinking too , as it is named on the internal working and not on the actual functionality/intent of the channel.

Long story short - Will pick the new name proposed.

File: index_alias_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index_alias_impl.go:58]: I am not sure the inconsistency(in terms of the number of indexes underneath) across calls is a problem here as that issue persists even today across search calls, just that it's hidden from app level?  But applications has to perform enough precautions here like freezing the plan of the index (index defn updates) etc during a forceMerge call.  And those higher level concerns needs to be handled separately was the thinking earlier..

> Another idea (just thinking out loud), what if instead the caller passed in a function, and we guaranteed to invoke that function on every index in the alias at that time. I'm not sure that's a whole lot better, but just wondering would that work for you?

Thinking out loud here on -- 
This would be slightly better with the above perceived problem in case we are holding a lock throughout the operation on all underlying indexes?
As the alias gets locked from further modifications during the potentially time consuming client callback executions perhaps its consistent that way...

Also, not sure whether we can come up with a generic, future ready callback signature here so that we won't regret tomorrow thinking that this is something the application has to take care and not at the bleve layer..

**Comment by mschoch** [comment]
Naming is hard for all of us.  In my experience often times the names made sense for an earlier version of the PR, and then as the PR evolves, the name still makes sense to the PR author (because you're sort of translating it internally), but it makes less sense to people reviewing it fresh.

The reality is that we all acknowledge that the concurrency in scorch is complicated, and to be honest this PR adds to that complexity.  That is why something like naming of the channels is even more important in this case.  Remember, we're all students of the @steveyen school of naming.

File: index/scorch/introducer.go [hunk 4]
```diff
@@ -313,11 +313,15 @@ func (s *Scorch) introduceMerge(nextMerge *segmentMerge) {
 
 	defer func() { _ = root.DecRef() }()
 
+	if nextMerge.creator == "" {
+		nextMerge.creator = "introduceMerge"
```
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index/scorch/introducer.go:317]: done!

File: index/scorch/merge.go [hunk 5]
```diff
@@ -106,6 +133,76 @@ OUTER:
...
+
+	if mr.MergeOptions == nil {
+		// assume the default single segment merge policy
+		mr.MergeOptions = &mergeplan.SingleSegmentMergePlanOptions
+	}
+	var ssCreator string
+	if mr.OverrideNumSnapshotsToKeep {
+		ssCreator = numSnapShotsToKeepOverRuler
+	}
+	msg := &mergerCtrl{creator: ssCreator,
```
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index/scorch/merge.go:185]: tried to fix it. lets see..

**Comment by sreekanth-cb** [comment]
@mschoch , this is ready for another round !

File: index_alias_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mschoch) [reply index_alias_impl.go:58]: Honestly, not sure the signature is a problem.  It should just be
```
func (i *indexAliasImpl) VisitIndexes(visit func(Index)) {
  i.mutex.RLock()
  defer i.mutex.RUnlock()
  for _, idx := range i.indexes {
    visit(idx)
  }
}
```

We only pas in the index, because that is the only thing the alias knows about.  If the caller needs access to something else, then they define their callback as a closure around that other data they need.

File: index/scorch/merge.go [hunk 6]
```diff
@@ -106,6 +134,58 @@ OUTER:
...
+	} else {
+		// assume the default single segment merge policy
+		mo = &mergeplan.SingleSegmentMergePlanOptions
+	}
+	msg := &mergerCtrl{options: mo,
+		doneCh: make(chan struct{}),
+		ctx:    ctx,
+	}
+
+	// kick the merger workloop
```
Comment (mschoch) [inline index/scorch/merge.go:172]: phrasing "kick" here still isn't friendly to other people reading the code later, how about:

// request the merger perform a force merge

File: index_alias_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index_alias_impl.go:58]: i see.. will switch to this, but i am failing to see any benefits here?

[the remark this answers is not in the document]
Comment (mschoch) [reply index_alias_impl.go:58]: We guarantee the index is still a part of the alias at the time the function is run.  The previous version cannot guarantee that.

File: index/scorch/merge.go (hunk 6 above)
Comment (sreekanth-cb) [reply index/scorch/merge.go:172]: done!

File: index_alias_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sreekanth-cb) [reply index_alias_impl.go:56]: done!
